controllers/replication_controller.go

```diff
@@ -132,78 +104,189 @@ func (r *ReconcileReplication) Reconcile(request reconcile.Request) (reconcile.R
 		err := r.assertExistingReplication(ctx, harborClient, replication)
 		if err != nil {
-			return reconcile.Result{}, err
+			return ctrl.Result{}, err
 		}
 
 	case registriesv1alpha1.ReplicationStatusPhaseTerminating:
 		// Delete the replication via harbor API
 		err := r.assertDeletedReplication(ctx, reqLogger, harborClient, replication)
 		if err != nil {
-			return reconcile.Result{}, err
+			return ctrl.Result{}, err
 		}
 	}
 
 	return r.updateReplicationCR(ctx, harbor, originalReplication, replication)
 }
 
+// reconcileRunningReplicationExecution fetches the newest replication execution of
+// the replication and checks, whether it is still running or not.
+// Returns an error, if no replication execution could be found for the replication policy.
+func (r *ReplicationReconciler) reconcileRunningReplicationExecution(ctx context.Context, replication *registriesv1alpha1.Replication, harborClient *h.RESTClient) (bool, error) {
+	replExec := &legacymodel.ReplicationExecution{
+		PolicyID: replication.Status.ID,
+		Trigger:  registriesv1alpha1.ReplicationTriggerTypeManual,
+	}
+
+	executions, err := harborClient.GetReplicationExecutions(ctx, replExec)
+	if err != nil {
+		return false, fmt.Errorf("no replication executions found: %s", err)
+	}
+
+	newestReplExecID := getNewestReplicationExecutionID(executions)
+
+	execution, err := harborClient.GetReplicationExecutionByID(ctx, newestReplExecID)
+	if err != nil {
+		return false, fmt.Errorf("could not get the latest running replication execution, id: %d, error: %s", newestReplExecID, err)
+	}
+
+	if execution.InProgress > 0 {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// reconcileFinishedReplicationExecution fetches the latest finished replication execution and returns an error when it has failed.
+func (r *ReplicationReconciler) reconcileFinishedReplicationExecution(ctx context.Context, replication *registriesv1alpha1.Replication, harborClient *h.RESTClient) error {
+	replExec := &legacymodel.ReplicationExecution{
+		PolicyID: replication.Status.ID,
+		Trigger:  registriesv1alpha1.ReplicationTriggerTypeManual,
+	}
+
+	executions, err := harborClient.GetReplicationExecutions(ctx, replExec)
+	if err != nil {
+		return fmt.Errorf("fetching successful replication executions failed for replication policy %s: %s", replication.Status.Name, err)
+	}
+
+	newestReplExecID := getNewestReplicationExecutionID(executions)
+
+	execution, err := harborClient.GetReplicationExecutionByID(ctx, newestReplExecID)
+	if err != nil {
+		return fmt.Errorf("could not get the latest finished replication execution, id: %d, error: %s", newestReplExecID, err)
+	}
+
+	if execution.Failed > 0 || execution.Status == "Failed" {
+		return fmt.Errorf("latest execution was not successful, replication policy: %s", replication.Spec.Name)
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReplicationReconciler implements reconcile.Reconciler.
+var _ reconcile.Reconciler = &ReplicationReconciler{}
```

Comment:
Should be moved to the top of the file. I didn't expect this "check" and the `SetupWithManager` between two reconciliation functions.